src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensEditsHandler.cs

```diff
@@ -0,0 +1,377 @@
+using Microsoft.VisualStudio.LanguageServer.Protocol;
+using Roslyn.Utilities;
+using LSP = Microsoft.VisualStudio.LanguageServer.Protocol;
+
+namespace Microsoft.CodeAnalysis.LanguageServer.Handler.SemanticTokens
+{
+    /// <summary>
+    /// Computes the semantic tokens edits for a file. An edit request is received every 500ms,
+    /// or every time an edit is made by the user.
+    /// </summary>
+    [ExportLspMethod(LSP.SemanticTokensMethods.TextDocumentSemanticTokensEditsName), Shared]
+    internal class SemanticTokensEditsHandler : AbstractRequestHandler<LSP.SemanticTokensEditsParams, SumType<LSP.SemanticTokens, LSP.SemanticTokensEdits>>
+    {
+        private readonly SemanticTokensCache _tokensCache;
+
+        [ImportingConstructor]
+        [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+        public SemanticTokensEditsHandler(
+            ILspSolutionProvider solutionProvider,
+            SemanticTokensCache tokensCache) : base(solutionProvider)
+        {
+            _tokensCache = tokensCache;
+        }
+
+        public override async Task<SumType<LSP.SemanticTokens, LSP.SemanticTokensEdits>> HandleRequestAsync(
+            LSP.SemanticTokensEditsParams request,
+            RequestContext context,
+            CancellationToken cancellationToken)
+        {
+            Contract.ThrowIfNull(request.TextDocument);
+
+            // Even though we want to ultimately pass edits back to LSP, we still need to compute all semantic tokens,
+            // both for caching purposes and in order to have a baseline comparison when computing the edits.
+            var newSemanticTokensData = await SemanticTokensHelpers.ComputeSemanticTokensDataAsync(
+                request.TextDocument, context.ClientName, SolutionProvider, SemanticTokensCache.TokenTypesToIndex,
+                range: null, cancellationToken).ConfigureAwait(false);
+
+            Contract.ThrowIfNull(newSemanticTokensData);
+
+            var resultId = _tokensCache.GetNextResultId();
+            var newSemanticTokens = new LSP.SemanticTokens { ResultId = resultId, Data = newSemanticTokensData };
+
+            await _tokensCache.UpdateCacheAsync(
+                request.TextDocument.Uri, newSemanticTokens, cancellationToken).ConfigureAwait(false);
+
+            // Getting the cached tokens for the document. If we don't have an applicable cached token set,
+            // we can't calculate edits, so we must return all semantic tokens instead.
+            var oldSemanticTokensData = await _tokensCache.GetCachedTokensDataAsync(
+                request.TextDocument.Uri, request.PreviousResultId, cancellationToken).ConfigureAwait(false);
+            if (oldSemanticTokensData == null)
+            {
+                return newSemanticTokens;
+            }
+
+            // The Data property is always populated on the server side, so it should never be null.
+            Contract.ThrowIfNull(oldSemanticTokensData);
```

Comment:
enh......... you literally check it for null one statement above.  seems really unnecessary.